ES1 and ES2 support

johnspicer opened the pull request:
@EDITD/hackers - This adds support for installing ES1 and ES2.

It install head plugin by default, but others can be specified. To keep support for ansible 1.x and 2.x, I'm not doing any `sudo`, or `become`, but the role expects that it has sudo privileges.

I have tested this in a playbook, with:

``` yaml
  - role: EDITD.elasticsearch_dev
    when: not production_mode
    become: yes
    become_method: sudo
    elasticsearch_version: 2.3.0
    elasticsearch_config_file: files/elasticsearch.dev.yml
```

Closes https://github.com/EDITD/ansible-elasticsearch_dev/issues/1

Commit:
Support both ES 1 and ES 2, using the ES deb filename, and add support for
specifying an ES config file

sujaymansingh commented on tasks/main.yml:
```diff
@@ -3,43 +3,39 @@
 - name: Install ES head plugin
-  command: /usr/share/elasticsearch/bin/plugin --install mobz/elasticsearch-head
-  sudo: yes
+  shell: /usr/share/elasticsearch/bin/plugin {% if elasticsearch_file_name|get_major_version == 1 %}--{% endif %}install {{ item }}
```
:+1:

johnspicer replied:
Ah this is a misleading name! https://github.com/EDITD/ansible-elasticsearch_dev/pull/3/commits/7b695ba90963afd44cabd6b0b00296b44c210e60

Commit:
Fix misleading ansible item name

sujaymansingh commented:
How come we don't check to see if ES is installed anymore? As it stands, if the deb file in `/tmp` isn't present, we will try to download and install even if ES is already installed. (And files in `/tmp` could/will get deleted!)

pkatseas commented on tasks/main.yml:
```diff
@@ -3,43 +3,39 @@
-  sudo: yes
-  when: elasticsearch_installed.stat.exists != true
+  get_url: url={{ elasticsearch_base_url }}/{{ elasticsearch_file_name }} dest=/tmp/{{ elasticsearch_file_name }}
+  when: elasticsearch_downloaded.stat.exists != true
```
This could be just `when: not elasticsearch_downloaded.stat.exists`

johnspicer replied:
👍  https://github.com/EDITD/ansible-elasticsearch_dev/pull/3/commits/5cf7ca617094ebc4169d82c718d885e293054260

pkatseas commented on tasks/main.yml:
```diff
@@ -3,43 +3,39 @@
+
+- name: write Elasticsearch config
+  copy: src={{ elasticsearch_config_file }} dest=/etc/elasticsearch/elasticsearch.yml
+  when: elasticsearch_config_file != false
```
Similarly here, `when: elasticsearch_config_file`

johnspicer replied:
I did this here https://github.com/EDITD/ansible-elasticsearch_dev/pull/3/commits/5cf7ca617094ebc4169d82c718d885e293054260, but undid it here https://github.com/EDITD/ansible-elasticsearch_dev/pull/3/commits/13c584aab7c6e70fb1a2c235b88d913b818f0fd5 after our convesation - it seems ansible is trying to evaluate the file path when checking!

pkatseas replied:
Yeah that looks ok.

pkatseas commented on tasks/main.yml:
```diff
@@ -3,43 +3,39 @@
 
-- name: Start ES
+- name: Make ES listen on all interfaces
+  lineinfile: 'dest=/etc/elasticsearch/elasticsearch.yml line="network.host: 0.0.0.0" regexp=^network.host'
```
In case there's already a line about `network.host` in a custom passed in `elasticsearch_config_file`, we'll be overwriting here. Should we add a default var to control this via a `when`, just in case?

johnspicer replied:
What do you think about this https://github.com/EDITD/ansible-elasticsearch_dev/pull/3/commits/97028d97ec29e49def7ba886a54f05a967c3943c? 

I couldn't think of a good default for this - either have to set it when in dev (quite annoying and would have to do it almost everywhere), or have to set it in production (not really wanted). Basically, i set it if we are deploying to a vagrant machine.

pkatseas replied:
@johnspicer Well I just had in mind a custom var that we'd default to true, like the `elasticsearch_config_file: false` one but along the lines of `elasticsearch_listen_on_all_interfaces: true`. What you're suggesting sounds good too, my only worry is that it might be a bit hard to understand why we're checking that and the value might change in a later version of Virtualbox or whatever.

johnspicer replied:
Yeah let's make it explicit! https://github.com/EDITD/ansible-elasticsearch_dev/pull/3/commits/cc4742bf97ee5599fab1f804c20107b31aeb06e1

johnspicer commented:
This is a good point @sujaymansingh. I was thinking I didn't like managing that through a file, just because if you switch ES versions then those files will still hang around (unless we delete them in the role as well). I'll look into doing it with something like [this](http://stackoverflow.com/questions/21892603/how-to-make-ansible-execute-a-shell-script-if-a-package-is-not-installed)

pkatseas commented:
I've left a couple of notes but it LGTM, nice one @johnspicer!

@sujaymansingh I think that relying on a file on disk is a bit flaky, if we want to be absolutely sure about ES being installed and the same version, why not check the actual ES version? For example:

``` bash
$ ls /usr/share/elasticsearch/lib/elasticsearch-*.jar
/usr/share/elasticsearch/lib/elasticsearch-1.7.3.jar
```

EDIT: Or @johnspicer's suggestion above (or anything that actually gives us the data we want)

sujaymansingh commented:
Checking the version installed sounds good to me!

Commit:
Simplify boolean checks in tasks

Commit:
The `!= false` seems to be needed here. Without it, ansible seems to be
trying to evaluate the file path, which fails.

Commit:
Check if ES is installed before installing it again

Commit:
Reinstall plugins in case they're already there. This isn't ideal, but
installing plugins is rapid, and the plugin install command fails angrily
if the plugin is already installed

Commit:
Only open ES to all interfaces if we are on vagrant

johnspicer commented:
@sujaymansingh - I now check whether ES is installed https://github.com/EDITD/ansible-elasticsearch_dev/pull/3/commits/a77e4cb910667d14bdfab12a0672ccf558da3bdc

sujaymansingh commented on tasks/main.yml:
```diff
@@ -25,7 +25,7 @@
 
 - name: Install ES plugins
-  shell: /usr/share/elasticsearch/bin/plugin {% if elasticsearch_file_name|get_major_version == 1 %}--{% endif %}install {{ item }}
+  shell: /usr/share/elasticsearch/bin/plugin {% if elasticsearch_file_name|get_major_version == 1 %}--{% endif %}remove {{ item }} && /usr/share/elasticsearch/bin/plugin {% if elasticsearch_file_name|get_major_version == 1 %}--{% endif %}install {{ item }}
```
Doesn't the `head_plugin_path` stop it from trying to install the plugin if it is already installed?

johnspicer replied:
It does, but as we support installing custom plugins, that are installed with a different name to the directory they are put in to, it means that we would need a check for each plugin and then a mapping from the name you use to install the plugin, and the directory that the plugin goes into.

The task above can be removed!

sujaymansingh replied:
Ah col.

johnspicer replied:
https://github.com/EDITD/ansible-elasticsearch_dev/pull/3/commits/0cc38897a2c3fb0026fb7887553c2822e3ac7e85

Commit:
Remove unneeded check - we just reinstall the plugins now

pkatseas commented on tasks/main.yml:
```diff
@@ -10,7 +10,7 @@
 - name: Download ES deb
   get_url: url={{ elasticsearch_base_url }}/{{ elasticsearch_file_name }} dest=/tmp/{{ elasticsearch_file_name }}
-  when: elasticsearch_downloaded.stat.exists != true
+  when: not elasticsearch_downloaded.stat.exists
```
:+1:

pkatseas commented on tasks/main.yml:
```diff
@@ -26,7 +26,7 @@
 - name: write Elasticsearch config
   copy: src={{ elasticsearch_config_file }} dest=/etc/elasticsearch/elasticsearch.yml
-  when: elasticsearch_config_file
+  when: elasticsearch_config_file != false
```
Booo ansible :-1:

pkatseas commented on tasks/main.yml:
```diff
@@ -8,12 +8,17 @@
+
 - name: Download ES deb
   get_url: url={{ elasticsearch_base_url }}/{{ elasticsearch_file_name }} dest=/tmp/{{ elasticsearch_file_name }}
   when: not elasticsearch_downloaded.stat.exists
```
Should this check for `elasticsearch_installed.stat.exists` too? No reason to re-download a deleted deb if we've already installed it.

johnspicer replied:
https://github.com/EDITD/ansible-elasticsearch_dev/pull/3/commits/01e6f0a9116f6257fadfb73d29ed271b4417e441

Commit:
Make explicit var for ES to listen on all interfaces

pkatseas commented on tasks/main.yml:
```diff
@@ -20,10 +20,6 @@
-
 - name: Install ES plugins
   shell: /usr/share/elasticsearch/bin/plugin {% if elasticsearch_file_name|get_major_version == 1 %}--{% endif %}remove {{ item }} && /usr/share/elasticsearch/bin/plugin {% if elasticsearch_file_name|get_major_version == 1 %}--{% endif %}install {{ item }}
   when: head_plugin_path.stat.exists != true
```
This should go too no?

johnspicer replied:
Yep! https://github.com/EDITD/ansible-elasticsearch_dev/pull/3/commits/011846f37e3f6e600fa9e0f032f238b7fbd893e0

pkatseas commented:
@johnspicer Oh also, could we please update the readme with all the extra bits and vars we have now? :pray:

sujaymansingh commented on tasks/main.yml:
```diff
@@ -31,7 +31,7 @@
 - name: Make ES listen on all interfaces
   lineinfile: 'dest=/etc/elasticsearch/elasticsearch.yml line="network.host: 0.0.0.0" regexp=^network.host'
-  when: ansible_product_name == "VirtualBox"
+  when: elasticsearch_listen_on_all_interfaces
```
:+1:

Commit:
Check that ES is installed before downloading it

Commit:
Don't reinstall the plugins, but just ignore the error if it's already
installed.

pkatseas commented on tasks/main.yml:
```diff
@@ -21,9 +21,9 @@
-  when: head_plugin_path.stat.exists != true
+  shell: /usr/share/elasticsearch/bin/plugin {% if elasticsearch_file_name|get_major_version == 1 %}--{% endif %}install {{ item }}
   with_items: "{{ plugins_to_install }}"
+  ignore_errors: True
```
This will ignore issues with trying to install an existing plugin, but I'm wary of it also ignoring errors we might want to know about, like legit "could not install the plugin" errors. Is there a way we can be more explicit about which errors we're ok with?

johnspicer replied:
The simplest way is probably

``` yaml
- name: Install ES plugins
  shell: /usr/share/elasticsearch/bin/plugin {% if elasticsearch_file_name|get_major_version == 1 %}--{% endif %}install {{ item }}
  with_items: "{{ plugins_to_install }}"
  register: plugin_response
  failed_when: "'ERROR' in plugin_response.stdout and 'uninstall it first' not in plugin_response.stdout"
```

'uninstall it first' is just a sample string that will almost certainly only be in that error. What do you think?

johnspicer replied:
I'll see if there's anything better in the output of that command though so we don't base it on a random string!

pkatseas replied:
Sounds good @johnspicer

johnspicer replied:
The error message is pretty short. 

```
ERROR: plugin directory /usr/share/elasticsearch/plugins/head already exists. To update the plugin, uninstall it first using 'remove head' command
```

Unfortunately we don't know the directory name so we can't use all of that. I don't think its that important to totally nail this down, plugins can just be installed if anything goes wrong. As long as we can run the playbook multiple times without it crashing I think its ok?

pkatseas replied:
@johnspicer I think your gist from above with either `uninstall it first` or `already exists` should suffice for this check

johnspicer replied:
Coooool beans https://github.com/EDITD/ansible-elasticsearch_dev/pull/3/commits/0203dc444106a71f8678a804fe534d66cc585bec

pkatseas replied:
Sweeet

Commit:
Update README with new options

Commit:
Only ignore the plugin install error if it's because the plugin is
already installed. Any other errors should still be errors

pkatseas commented:
Latest changes look great @johnspicer !

johnspicer commented:
Thanks for the reviews @pkatseas and @sujaymansingh!
